### PR TITLE
Fix building the Dub frontend package on Windows

### PR DESCRIPTION
**Pull request by jacob-carlborg**
* The Dub package was unconditionally running the `config.sh` shell script to generate the `VERSION` file. A Bash shell script will not work on Windows. Therefore the `config.sh` script has been ported to D, which will work on both Windows and Posix
* The Dub package tests has been fixed so they work on Windows
* The Dub package tests are now being run by AppVeyor

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Review by wilzbach — APPROVED** [review]
Not a Bat expert, but looks to do the same. Could you maybe add the dub build to appveyor so that we have a CI which can verify this?
Also, it probably would be helpful to copy the comments from config.sh over and add notes in both files that they should be kept in sync.

**Comment by wilzbach** [comment]
Wait, before we commit to such duplications: why not convert the script to D? A host compiler is already required.

**Comment by jacob-carlborg** [comment]
> Wait, before we commit to such duplications: why not convert the script to D? A host compiler is already required.

They're not the same, but I guess I can version that out.

**Comment by jacob-carlborg** [comment]
So far Phobos isn't used in the compiler. Can I use Phobos for this?

**Comment by wilzbach** [comment]
> So far Phobos isn't used in the compiler. Can I use Phobos for this?

Yes of course. This is just a build script. The other build scripts also depend on Phobos.

**Comment by jacob-carlborg** [comment]
I don't understand why the builds are failing, it works locally and the auto-tester passes.

**Comment by thewilsonator** [comment]
```
dmd/globals.d:366:21: error: file "VERSION" cannot be found or not in a path specified with -J
         _version = (import("VERSION") ~ '\0').ptr;
```

**Comment by jacob-carlborg** [comment]
It might be due to I had `@safe:` at the top. And not all those functions the script is calling are safe in older versions of Phobos.

**Comment by jacob-carlborg** [comment]
Still the same issue for buildkite:

```
*** No rule to make target '../generated/linux/release/64/VERSION', needed by '../generated/linux/release/64/dmd'. Stop.
```.

**Comment by wilzbach** [comment]
No you need to depend on the host compiler for buildkite and DAutotest as they do download the bootstrapping compiler. I can add it an half an hour or so.

**Comment by wilzbach** [comment]
Should be passing Buildkite now (though my internet atm is too slow to test auto-bootstrapping myself)

File: config.d [hunk 1]
```diff
@@ -0,0 +1,69 @@
+void main(const string[] args)
```
Comment (wilzbach) [inline config.d]: Maybe a short header of what this script does and that its executed as part of the build process to update the VERSION file?

**Review by wilzbach — APPROVED** [review]

File: src/posix.mak [hunk 2]
```diff
@@ -544,8 +544,21 @@ $(optabgen_files): optabgen.out
...
+# actually has changed.
+
+$G/version_check: ../config.d $(HOST_DMD_RUN)
+	@echo "  (HOST_DMD_RUN)  $<  $<"
+	$(HOST_DMD_RUN) $< -of$@
+
+$G/VERSION: $G/version_check ../VERSION FORCE
+	@$< $G ../VERSION $(SYSCONFDIR)
+
+$G/SYSCONFDIR.imp: $G/VERSION
```
Comment (wilzbach) [inline src/posix.mak]: It probably would be nicer to do this directly in this target (would also allow us to drop the special casing from `config.d`:

```diff
diff --git a/config.d b/config.d
index 606a2b62c..a214ad9b4 100644
--- a/config.d
+++ b/config.d
@@ -6,19 +6,10 @@ void main(const string[] args)
     const outputDirectory = args[1];
     const versionFile = args[2];
 
-    version (Posix)
-        const sysConfigDirectory = args[3];
-
     mkdirRecurse(outputDirectory);
     const version_ = generateVersion(versionFile);
 
     updateIfChanged(buildPath(outputDirectory, "VERSION"), version_);
-
-    version (Posix)
-    {
-        const path = buildPath(outputDirectory, "SYSCONFDIR.imp");
-        updateIfChanged(path, sysConfigDirectory);
-    }
 }
 
 /**
diff --git a/src/posix.mak b/src/posix.mak
index 3a83236eb..724e95285 100644
--- a/src/posix.mak
+++ b/src/posix.mak
@@ -554,9 +554,10 @@ $G/version_check: ../config.d $(HOST_DMD_RUN)
 	$(HOST_DMD_RUN) $< -of$@
 
 $G/VERSION: $G/version_check ../VERSION FORCE
-	@$< $G ../VERSION $(SYSCONFDIR)
+	@$< $G ../VERSION
 
-$G/SYSCONFDIR.imp: $G/VERSION
+$G/SYSCONFDIR.imp:
+	@[ "$(cat $@)" == "$(SYSCONFDIR)" ] || echo "$(SYSCONFDIR)" > $@
 
 FORCE: ;
```

File: dub.sdl [hunk 3]
```diff
@@ -29,7 +29,8 @@ subPackage {
     "src/dmd/tokens.d" \
     "src/dmd/utf.d"
 
-  preGenerateCommands `cd "$${DUB_PACKAGE_DIR}" && ./config.sh generated/dub VERSION /etc`
+  preGenerateCommands `cd "$${DUB_PACKAGE_DIR}" && $${DC} -run ./config.d generated/dub VERSION /etc` platform="posix"
```
Comment (wilzbach) [inline dub.sdl]: DAutoTest fails with:

```
.generated/stable_dmd-2.083.0-beta.2/dmd2/linux/bin64/dub build --compiler=.generated/stable_dmd-2.083.0-beta.2/dmd2/linux/bin64/dmd --root=ddoc && \
	mv ddoc/ddoc_preprocessor .generated/ddoc_preprocessor
Fetching libdparse 0.8.7 (getting selected version)...
Fetching stdx-allocator 2.77.2 (getting selected version)...
Invalid source/import path: /dev/shm/dtest/work/repo/dmd/generated/dub
Running pre-generate commands for dmd:frontend...
/bin/sh: .generated/stable_dmd-2.083.0-beta.2/dmd2/linux/bin64/dmd: No such file or directory
Command failed with exit code 127: cd "${DUB_PACKAGE_DIR}" && ${DC} -run ./config.d generated/dub VERSION /etc
posix.mak:922: recipe for target '.generated/ddoc_preprocessor' failed
```

So it looks like you need to save PWD

Comment (wilzbach) [reply dub.sdl]: (did this for you)

File: config.d (hunk 1 above)
Comment (jacob-carlborg) [reply config.d]: Done.

File: dub.sdl (hunk 3 above)
Comment (jacob-carlborg) [reply dub.sdl]: 👍.

File: src/posix.mak (hunk 2 above)
Comment (jacob-carlborg) [reply src/posix.mak]: This did not work. The `SYSCONFDIR.imp` is needed by the `dinifile.d` file which is included in the Dub package. I added back the code as it was.

File: test/dub_package/frontend_file.d [hunk 4]
```diff
@@ -42,14 +43,19 @@ void main()
...
 
     t.module_.fullSemantic;
-    auto generated = t.module_.prettyPrint;
+    auto generated = t.module_.prettyPrint.toUnixLineEndings();
 
-    auto expected =q{module foo;
+    version (Windows)
+        enum accumulator = "0.000000";
+    else
+        enum accumulator = "0.00000";
```
Comment (wilzbach) [inline test/dub_package/frontend_file.d]: Why is this necessary?

File: test_dub_package_impl.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
+test_dub_package_impl() {
```
Comment (wilzbach) [inline test_dub_package_impl.sh]: `set -euxo pipefail`

**Review by wilzbach — COMMENTED** [review]
It would have been fine for me to do Appveyor in another PR.

(Btw I won't have much time for the entire month)

File: test/dub_package/frontend_file.d (hunk 4 above)
Comment (jacob-carlborg) [reply test/dub_package/frontend_file.d]: I don't know. The generated code is different on Windows and Posix. Something related to the C standard library that is probably used to format the floating point number?

File: test_dub_package_impl.sh (hunk 5 above)
Comment (jacob-carlborg) [reply test_dub_package_impl.sh]: This is file is intended to be sourced in other files.

File: ci.sh [hunk 6]
```diff
@@ -95,22 +95,7 @@ test_dmd() {
...
-            # build with host compiler
-            dub --single "$file"
-            # build with built compiler (~master)
-            DFLAGS="-de" dub --single --compiler="${abs_build_path}/dmd" "$file"
-        done
-        popd
-        # Test rdmd build
-        "${build_path}/dmd" -version=NoBackend -version=GC -version=NoMain -Jgenerated/dub -Jres -Isrc -i -run test/dub_package/frontend.d
-    fi
+    test_dub_package_impl
```
[the remark this answers is not in the document]
Comment (jacob-carlborg) [reply ci.sh]: It doesn't matter where it's sourced, since the sourced script contains a function. The function needs to be called anyway.

[the remark this answers is not in the document]
Comment (wilzbach) [reply ci.sh]: Yes, but it matters whether you source or call it. Calling is imho better because it avoids any potential issues with the namespace and allows to reproduce it easier.

File: test/dub_package/frontend_file.d (hunk 4 above)
Comment (wilzbach) [reply test/dub_package/frontend_file.d]: Would be nice to have this as a comment.

Comment (jacob-carlborg) [reply test/dub_package/frontend_file.d]: Sure.

File: ci.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (jacob-carlborg) [reply ci.sh]: Ah, you mean calling the whole script? I can do that.

File: test/dub_package/frontend_file.d (hunk 4 above)
Comment (jacob-carlborg) [reply test/dub_package/frontend_file.d]: Fixed.

**Comment by jacob-carlborg** [comment]
> It would have been fine for me to do Appveyor in another PR.

Removed the AppVeyor tests.

File: test_dub_package_impl.sh (hunk 5 above)
Comment (jacob-carlborg) [reply test_dub_package_impl.sh]: Not relevant anymore.

**Comment by jacob-carlborg** [comment]
All green (except semaphoreci, which I'm guessing is unrelated). @thewilsonator

**Comment by jacob-carlborg** [comment]
Thanks.
